### PR TITLE
ICU-21339 toSkeleton: treat percent and permille correctly

### PR DESCRIPTION
**Pull request by hugovdm**
##### Checklist

- [X] Issue filed: https://unicode-org.atlassian.net/browse/ICU-21339
- [X] Updated PR title and link in previous line to include Issue number
- [x] Issue accepted
- [X] Tests included
- [ ] Documentation is changed or added

### CODE REVIEW

**Review by sffc — DISMISSED** [review]

File: icu4c/source/i18n/number_skeletons.cpp
Comment (FrankYFTang) [inline icu4c/source/i18n/number_skeletons.cpp:1505]: Isn't the !utils::unitIsBaseUnit(macros.perUnit) && utils::unitIsCurrency(macros.unit)) => U_UNSUPPORTED_ERROR / false
a new behavior?

**Review by FrankYFTang — COMMENTED** [review]
Since you are changing the code around currency, could you add unit test to test the condition
1. check the currency -per- meter
2. check meter -per- currency
Also, see does the expectation changed if you run that on 67. IF so, is the change reasonable?

File: icu4c/source/i18n/number_skeletons.cpp
Comment (hugovdm) [reply icu4c/source/i18n/number_skeletons.cpp:1505]: You're right that I've hereby changed the toSkeleton output. I'm not sure what the best thing to do is - is it just a matter of needing to document the behaviour change?

When constructing a NumberFormatter, what happens (in number_formatimpl.cpp, macrosToMicroGenerator FWIW) is if unit is set to currency, perUnit was simply ignored in 67, and is still ignored in 68. Is it better to preserve this behaviour, simply ignoring perUnit?

Comment (hugovdm) [reply icu4c/source/i18n/number_skeletons.cpp:1505]: I'll add the suggested unit tests. Thoughts on the question above ("is the change reasonable?"):

- It is indeed a behaviour change. I do think the change is reasonable, though arguably we should give an error already at the formatter construction, not only at the toSkeleton() call... and I may have given insufficient thought to what we might end up breaking with this? Do you think it's risky?

- For the remaining part of ICU-20941 for ICU 69, we *could* consider adding support for $CURRENCY-per-meter, but this doesn't fall into the unit identifier spec, so we may need to carefully consider the right way to deal with this case (e.g. also in the specs).

Comment (sffc) [reply icu4c/source/i18n/number_skeletons.cpp:1505]: We already announced that skeleton behavior for units is changing in ICU 68.  The RC version of that behavior was buggy, as Frank found.  I think it is 110% in scope to fix behavior between RC and GA according to requests from clients like Frank.

**Review by sffc — DISMISSED** [review]

File: icu4c/source/i18n/number_skeletons.cpp
Comment (hugovdm) [reply icu4c/source/i18n/number_skeletons.cpp:1505]: Turns out forSkeleton doesn't support currency: blueprint_helpers::parseMeasureUnitOption returns an U_INTERNAL_PROGRAM_ERROR because there are more than 40 currencies - it supports only unit types with up to 40 subtypes.

The new toSkeleton behaviour can still be triggered by constructing a NumberFormatter with the fluent API.

Comment (sffc) [reply icu4c/source/i18n/number_skeletons.cpp:1505]: "currency-per-meter" requires more work, not only in skeletons but also in the formatting step.

I like the currently proposed fix because it fixes "percent-per-meter", for which we have formatting data, without touching compound currency units, for which we do not have formatting data.

Comment (hugovdm) [reply icu4c/source/i18n/number_skeletons.cpp:1506]: Changes made:
- for currencies, reject `perUnit` at `forSkeleton` time, just as `unit` is rejected.
- unit test demonstrating toSkeleton fails when fluent-API is used to construct a currency/unit conflict.

For ICU 69, we may need to revisit what is produced by the formatter: possibly an error, rather than ignoring perUnit, or possibly supporting it, after figuring out how the skeletons will be handled for currency-vs-unit.

File: icu4j/main/classes/core/src/com/ibm/icu/number/NumberSkeletonImpl.java
Comment (younies) [inline icu4j/main/classes/core/src/com/ibm/icu/number/NumberSkeletonImpl.java:1499]: unit.equals(MeasureUnit.PERCENT)

File: icu4j/main/classes/core/src/com/ibm/icu/number/NumberSkeletonImpl.java
Comment (younies) [inline icu4j/main/classes/core/src/com/ibm/icu/number/NumberSkeletonImpl.java:1502]: unit.equals(MeasureUnit. PERMILLE)

**Review by younies — CHANGES_REQUESTED** [review]

File: icu4j/main/classes/core/src/com/ibm/icu/number/NumberSkeletonImpl.java
Comment (hugovdm) [reply icu4j/main/classes/core/src/com/ibm/icu/number/NumberSkeletonImpl.java]: Done. Thanks!

File: icu4j/main/classes/core/src/com/ibm/icu/number/NumberSkeletonImpl.java
Comment (hugovdm) [reply icu4j/main/classes/core/src/com/ibm/icu/number/NumberSkeletonImpl.java]: Done.

File: icu4c/source/i18n/number_skeletons.cpp
Comment (sffc) [inline icu4c/source/i18n/number_skeletons.cpp:735]: Can you make the equivalent change in Java?

**Review by sffc — CHANGES_REQUESTED** [review]

File: icu4c/source/i18n/number_skeletons.cpp
Comment (FrankYFTang) [reply icu4c/source/i18n/number_skeletons.cpp:1506]: I have no strong opinion about CURRENCY-per-other unit as well as 
1. you add a unit test for that
2. you understand the impact of the changes 
3. V8 will not hit that case and it will not impact the v8/chrome integration work.
I will leave it to you and shane to decide is the changed behavior is acceptable to 68 or not. But we need to have Unit test to test whatever behavior you and shane agree to.

File: icu4c/source/i18n/number_skeletons.cpp
Comment (hugovdm) [reply icu4c/source/i18n/number_skeletons.cpp:735]: Done. Required a slightly different implementation because CHECK_NULL in C++ sets the seen flag, whereas in Java checkNull only checks for references.

**Review by FrankYFTang — DISMISSED** [review]
Thanks you for fixing this promptly

**Comment by hugovdm** [comment]
Squashed. The commit that @sffc has not reviewed yet: b837dad (in case he wants to see).

**Review by younies — APPROVED** [review]

**Comment by hugovdm** [comment]
I'm waiting a bit longer to see if @FrankYFTang wants to rubber-stamp the squash (as I have pre-squash approval), and/or whether @sffc wants to review the final changes. I'm otherwise considering merging when the ICU meeting starts.

**Review by sffc — APPROVED** [review]

**Pull request merged**
